File: questions/20_valid_parentheses/code.js
/**
 * @param {string} s
 * @return {boolean}
 */
var isValid = function(s) {
    const stack = [];
const map = {
    '(': ')',
    '{': '}',
    '[': ']'
};

for (let i = 0; i < s.length; i++) {
    const current = s[i];

    // only opening brackets are keys in the map object, so if it's not a key, it must be a closing
    if (map[current]) {
        // If it's an opening bracket, push to the stack
        stack.push(current);
    } else {
        // If it's a closing bracket, check the top of the stack
        const last = stack.pop();
        if (map[last] !== current) {
            return false;
        }
    }
}
return stack.length === 0;
};

File: questions/20_valid_parentheses/code.go
func isValid(s string) bool {
    stack := []rune{} // Stack to store opening brackets

    for _, char := range s {
        if char == '(' || char == '{' || char == '[' {
            stack = append(stack, char) // Push to stack
        } else {
            if len(stack) == 0 {
                return false // No opening bracket to match
            }
            top := stack[len(stack)-1] // Get the top element
            stack = stack[:len(stack)-1] // Pop from stack

            if (top == '(' && char != ')') || 
               (top == '{' && char != '}') || 
               (top == '[' && char != ']') {
                return false // Mismatched pair
            }
        }
    }

    return len(stack) == 0 // If stack is empty, it's valid
}